### PR TITLE
[FIRRTL][NFC] Add portAnnotations to FModule and FExtModule

### PR DESCRIPTION
**Pull request by darthscsi**
Add an explicit attribute for port annotations to FModule and FExtModule.  Adding a common name for port annotations, similar to instance, mem, etc, makes manipulating annotations simpler.  Although this could be folded into argattr, having it in the same format as used in other ops makes annotation scatting able to be operation type invariant, thus simplifying it.

### CODE REVIEW

**Comment by darthscsi** [comment]
This is in anticipation of the annotation scattering rework.

**Review by seldridge — COMMENTED** [review]
I think this change would also require an update to the `forPort` methods on `AnnotationSet`.  (Those use `mlir::function_like_impl::getArgAttrs` which would need to be change.) 

I'm surprised that the tests are passing with this patch, though. 🤔

File: include/circt/Dialect/FIRRTL/FIRRTLTypes.td [hunk 1]
```diff
@@ -178,8 +178,19 @@ def SubAnnotationAttr
...
+def PortAnnotationsAttr : ArrayAttrBase<
+    And<[
+      // Guarantee this is an ArrayAttr first
+      CPred<"$_self.isa<::mlir::ArrayAttr>()">,
+      // Guarantee all elements are DictionaryAttr or SubAnnotationAttr
+      CPred<"::llvm::all_of($_self.cast<::mlir::ArrayAttr>(), "
+            "[&](::mlir::Attribute attr) { return attr.isa<"
+            "::mlir::ArrayAttr,"
+            "::mlir::DictionaryAttr>();})">]>,
+    ""> {
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLTypes.td:190]: These explicit constraints look great!

**Review by seldridge — APPROVED** [review]

**Review by lattner — APPROVED** [review]
nice

**Pull request merged**